File: solutions/6/main.py
from typing import List 

class Solution:
    # ref https://leetcode-cn.com/problems/zigzag-conversion/solution/zzi-xing-bian-huan-by-jyd/
    def convert(self, s: str, numRows: int) -> str:
        # check input
        if numRows == 1:
            return s
        result = ["" for _ in range(numRows)]
        i = 0
        flag = -1
        for c in s:
            result[i] += c
            if i == 0 or i == numRows-1:
                flag *= -1
            i += flag
        return "".join(result)

if __name__ == "__main__": 
    s = "PAYPALISHIRING"
    numRows = 3
    ans = "PAHNAPLSIIGYIR"
    sol = Solution()
    assert sol.convert(s, numRows) == ans 

File: solutions/6/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sun Nov 28 13:12:46 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <string> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    string convert(string s, int numRows) {
        // check input
        if (numRows == 1) return s;
        vector<string> result(numRows, "");
        int i = 0;
        int flag = -1;
        for (auto& c : s) {
            result[i] += c;
            if (i == 0 || i == numRows-1) flag = -flag;
            i += flag;
        }
        string res;
        for (auto& r : result) res += r;
        return res;
    }
};

TEST(Test6, SimpleTest) {
    string s = "PAYPALISHIRING"; 
    const int numRows = 3;
    string ans = "PAHNAPLSIIGYIR";
    Solution sol; 
    EXPECT_EQ(sol.convert(s, numRows), ans);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
